gemrb/core/Scriptable/Scriptable.cpp

```diff
@@ -2174,51 +2167,152 @@ unsigned char Movable::GetNextFace()
+							oldPos = Pos;
+							bumped = false;
+							bumpBackTries = 0;
+							if (SquaredDistance(Pos, Destination) < unsigned(size * 32 * size * 32)) {
+								ClearPath(true);
+							}
+						}
+					}
+				}
+
+				return;
+			}
+		}
+		bumped = false;
+		MoveTo(oldPos);
+		bumpBackTries = 0;
+		return;
 	}
 	if (!time) time = core->GetGame()->Ticks;
-	if (!walk_speed) {
+	if (!walkScale) {
 		// zero speed: no movement
-		timeStartStep = time;
 		StanceID = IE_ANI_READY;
-		return true;
+		timeStartStep = time;
+		return;
 	}
 	if (!step) {
 		step = path;
 		timeStartStep = time;
-	} else if (step->Next && (( time - timeStartStep ) >= walk_speed)) {
-		//print("[New Step] : Orientation = %d", step->orient);
-		step = step->Next;
-		timeStartStep = timeStartStep + walk_speed;
-	}
-	SetOrientation (step->orient, true);
-	StanceID = IE_ANI_WALK;
-	if ((Type == ST_ACTOR) && (InternalFlags & IF_RUNNING)) {
-		StanceID = IE_ANI_RUN;
-	}
-	Pos.x = ( step->x * 16 ) + 8;
-	Pos.y = ( step->y * 12 ) + 6;
-	if (!step->Next) {
-		// we reached our destination, we are done
-		ClearPath();
-		NewOrientation = Orientation;
-		//since clearpath no longer sets currentaction to NULL
-		//we set it here
-		//no we don't, action is responsible for releasing itself
-		//ReleaseCurrentAction();
-		return true;
+		return;
 	}
-	if (( time - timeStartStep ) >= walk_speed) {
-		// we didn't finish all pending steps, yet
-		return false;
+
+	Point nmptStep(step->x, step->y);
+	double dx = nmptStep.x - Pos.x;
+	double dy = nmptStep.y - Pos.y;
+	Map::NormalizeDeltas(dx, dy, double(gamedata->GetStepTime()) / double(walkScale));
+	if (time > timeStartStep) {
+		Actor *actorInTheWay = nullptr;
+		// We can't use GetActorInRadius because we want to only check directly along the way
+		// and not be blocked by actors who are on the sides
+		int collisionLookaheadRadius = ((size < 3 ? 3 : size) - 1) * 3;
+		for (int r = collisionLookaheadRadius; r > 0 && !actorInTheWay; r--) {
+			double xCollision = Pos.x + dx * r;
+			double yCollision = Pos.y + dy * r * 0.75;
+			Point nmptCollision(xCollision, yCollision);
+			actorInTheWay = area->GetActor(nmptCollision, GA_NO_DEAD|GA_NO_UNSCHEDULED);
+		}
+		if (BlocksSearchMap() && actorInTheWay && actorInTheWay != this && actorInTheWay->BlocksSearchMap()) {
+			// Give up instead of bumping if you are close to the goal
+			if (!(step->Next) && std::abs(nmptStep.x - Pos.x) < XEPS * 3 && std::abs(nmptStep.y - Pos.y) < YEPS * 3) {
+				ClearPath(true);
+				NewOrientation = Orientation;
+				return;
+			}
+			if (Type == ST_ACTOR && ((Actor*)this)->ValidTarget(GA_CAN_BUMP) && actorInTheWay->ValidTarget(GA_ONLY_BUMPABLE)) {
+				actorInTheWay->BumpAway();
+			} else {
+				Backoff();
+				return;
+			}
+		}
+		// Stop if there's a door in the way
+		if (BlocksSearchMap() && area->GetBlockedNavmap(Pos.x + dx, Pos.y + dy) & PATH_MAP_SIDEWALL) {
+			Log(DEBUG, "DoStep", "%s stopping because of a door", GetName(0));
```

Comment:
I think this is one of the noisy ones. Valuable, remove later or remove now?

Reply:
Remove now IMO

Reply:
It's especially noisy with random walk, which leaves the actor to stop on this call when it encounters a wall